### PR TITLE
implement `.isEmpty()` method

### LINKED ISSUE #1356: Request for `isEmpty()` Set operator
### Category

Cedar language or syntax features/changes

### Describe the feature you'd like to request

The request is to add an `isEmpty()` operator that when applied to `Set`s, returns a boolean describing whether the set is empty or not.  This should be a valid operator usage even when validation is enabled.  The following would be a sample usage:

```cedar
permit (
  principal,
  action == Action::"claimPrize",
  resource
) when { context.claimedUsers.isEmpty() }
```

Today, if validation is not enabled, we can check for set emptiness with `context.mySet == []`.  However, [according to the docs](https://docs.cedarpolicy.com/policies/syntax-datatypes.html#datatype-set), `[]` is not considered a valid Set.  So, this syntax as is will not work when validation is enabled.

There exists a workaround today (credit to Craig Disselkoen) for when validation is enabled.

> The problem is that the validator needs to know the type of [], which needs to be the same as the type of context.claimedUsers, and Cedar doesn’t have type inference (yet).  The workaround for that is to declare some additional context attribute like emptyUsers of the appropriate type (the same type as claimedUsers), and pass the empty-set as the value for that context attribute, and then in your policy you can write context.claimedUsers == context.emptyUsers.

It would be great if users did not need to add workaround attributes to their contexts to mimic this state -- hence, an `isEmpty()` operator.

### Describe alternatives you've considered

There exists a workaround today in the event that validation is enabled, but it can be inconvenient to define a separate attribute per type of Set.  The workaround is copied above.

### Additional context

_No response_

### Is this something that you'd be interested in working on?

- [X] 👋 I may be able to implement this feature request
- [ ] ⚠️ This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by cdisselkoen**
## Description of changes

New builtin `.isEmpty()` method, which works on Sets.  Implemented similarly to `.contains()` which works on Sets.  This is our first builtin method that takes 0 arguments other than its receiver (although we have extension methods that take 0 arguments other than their receiver).

This PR contains the changes necessary in this repo (unless I've missed something); we'll also need changes in DRT, the Lean model, and the public docs.

## Issue #, if available

Fixes #1356 

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

I confirm that [`docs.cedarpolicy.com`](https://docs.cedarpolicy.com/) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in [`cedar-docs`](https://github.com/cedar-policy/cedar-docs). PRs should be targeted at a `staging-X.Y` branch, not `main`.)

### CODE REVIEW

File: cedar-policy-validator/src/entity_manifest.rs [hunk 1]
```diff
@@ -538,8 +538,19 @@ fn entity_manifest_from_expr(
...
-                // both unary ops are on booleans, so they are simple
+                // these unary ops are on primitive types, so they are simple
                 UnaryOp::Not | UnaryOp::Neg => Ok(entity_manifest_from_expr(arg)?.empty_paths()),
+                UnaryOp::IsEmpty => {
+                    // PANIC SAFETY: Typechecking succeeded, so type annotations are present.
+                    #[allow(clippy::expect_used)]
+                    let ty = arg
+                        .data()
+                        .as_ref()
+                        .expect("Expected annotated types after typechecking");
```
Comment (adpaco-aws) [inline cedar-policy-validator/src/entity_manifest.rs:549]: I don't understand why we do this - it's just a sanity check?

File: cedar-policy-core/src/ast/ops.rs [hunk 2]
```diff
@@ -32,6 +32,20 @@ pub enum UnaryOp {
...
+    ///
+    /// Argument must have Set type
+    IsEmpty,
+}
+
+impl std::fmt::Display for UnaryOp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            UnaryOp::Not => write!(f, "!_"),
+            UnaryOp::Neg => write!(f, "-_"),
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-core/src/ast/ops.rs:45]: Good question. Not sure why we even have this `Display` impl at all.

File: cedar-policy-validator/src/entity_manifest.rs (hunk 1 above)
Comment (cdisselkoen) [reply cedar-policy-validator/src/entity_manifest.rs:549]: I don't understand any of this `entity_manifest.rs` code.  I was trying to take the relevant code from `BinaryApp` and adapt it in a sane way for `UnaryApp`

Comment (john-h-kastner-aws) [reply cedar-policy-validator/src/entity_manifest.rs:549]: I don't understand either, but this looks reasonable to me

Comment (adpaco-aws) [reply cedar-policy-validator/src/entity_manifest.rs:549]: Yeah, looks good assuming the `BinaryApp` case is fine.

File: cedar-policy-core/src/ast/ops.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adpaco-aws) [reply cedar-policy-core/src/ast/ops.rs:45]: Then I think we should delete it.

File: cedar-policy-validator/src/typecheck.rs [hunk 3]
```diff
@@ -2390,6 +2390,23 @@ impl<'a> Typechecker<'a> {
...
                 })
             }
+            UnaryOp::IsEmpty => {
+                let ans_arg = self.expect_type(
+                    request_env,
+                    prior_capability,
+                    arg,
+                    Type::any_set(),
+                    type_errors,
+                    |_| None,
```
[the remark this answers is not in the document]
Comment (john-h-kastner-aws) [reply cedar-policy-validator/src/typecheck.rs]: probably not for records actually. `context == {}` will only ever work if `context` is declared without _any_ attributes due to #1303

File: cedar-policy-core/src/ast/ops.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-core/src/ast/ops.rs:45]: Tried to delete and found that it is indeed used in some error messages (e.g., in `InvalidRestrictedExpressionError` in restricted_expr.rs).  But the `_` doesn't make sense in that context; removing it

[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-core/src/ast/ops.rs]: Removed.

File: cedar-policy-validator/src/typecheck.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/typecheck.rs]: Fair, I'll remove the record case

File: cedar-policy-validator/src/typecheck.rs [hunk 4]
```diff
@@ -2390,6 +2390,23 @@ impl<'a> Typechecker<'a> {
                     )
                 })
             }
+            UnaryOp::IsEmpty => {
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/typecheck.rs:2393]: Added two tests

**Review by john-h-kastner-aws — APPROVED** [review]

**Pull request merged**